File: 2713-maximum-strictly-increasing-cells-in-a-matrix/2713-maximum-strictly-increasing-cells-in-a-matrix.py
class Solution:
    def maxIncreasingCells(self, mat: List[List[int]]) -> int:
        dic = defaultdict(list)
        m = len(mat)
        n = len(mat[0])
        
        for row in range(m):
            for col in range(n):
                dic[mat[row][col]].append((row, col))
        
        rows = [0] * m
        cols = [0] * n
        
        for _, arr in sorted(dic.items()):
            curr = []
            for r, c in arr:
                curr.append(1 + max(rows[r], cols[c]))
            
            for i in range(len(arr)):
                r, c = arr[i]
                x = curr[i]
                rows[r] = max(rows[r], x)
                cols[c] = max(cols[c], x)
            
        return max(rows + cols)

File: 2713-maximum-strictly-increasing-cells-in-a-matrix/2713-maximum-strictly-increasing-cells-in-a-matrix.js
const stmkey_in = (m) => new Map([...m].sort((x, y) => x[0] - y[0]));

const maxIncreasingCells = (g) => {
    let n = g.length, m = g[0].length, ma = new Map(),
        row = new Int32Array(n).fill(0), col = new Int32Array(m).fill(0), res = 0;
    for (let i = 0; i < n; i++) {
        for (let j = 0; j < m; j++) {
            if (!ma.has(g[i][j])) ma.set(g[i][j], []);
            ma.get(g[i][j]).push([i, j]);
        }
    }
    ma = stmkey_in(ma);
    for (const [x, d] of ma) {
        let trow = [], tcol = [];
        for (const [i, j] of d) {
            let t = Math.max(1, row[i] + 1, col[j] + 1);
            res = Math.max(res, t);
            trow.push([i, t]);
            tcol.push([j, t]);
        }
        for (const [i, j] of trow) row[i] = Math.max(row[i], j);
        for (const [i, j] of tcol) col[i] = Math.max(col[i], j);
    }
    return res;
};